# Problem 2

## Data Import
First, download the movielens (small) [dataset](https://files.grouplens.org/datasets/movielens/ml-latest-small.zip) as `pandas.DataFrame` objects. 

In [1]:
import pandas as pd

path = "Misc_files/movielens_data/ml-latest-small/"

# load movies and ratings DataFrames
movies = pd.read_csv(path+"movies.csv", header=0)
ratings = pd.read_csv(path+"ratings.csv", header=0)

We can then use the `head()` method to see the raw format of these `DataFrame` objects.

In [2]:
n_movies = len(movies)

print(f"Number of Unique Movies: {n_movies}")
movies.head()

Number of Unique Movies: 9742


movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [3]:
n_ratings = len(movies)
n_users = ratings.userId.nunique()
n_rated_movies = ratings.movieId.nunique()

print(f"Number of Ratings: {n_ratings}\nNumber of Users: {n_users}\nNumber of Unique Rated Movies: {n_rated_movies}")
ratings.head()

Number of Ratings: 9742
Number of Users: 610
Number of Unique Rated Movies: 9724


userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

Upon inspection of the raw data we note that of the 9,742 movies in the `movies` DataFrame, only 9,724 movies have been rated.

## Preprocessing
We will be implementing the steps found in this [Princeton](https://www.cs.princeton.edu) _Movie Embeddings_ [problem](https://www.cs.princeton.edu/courses/archive/fall16/cos402/ex/MovieEmbedding.pdf).

### Co-occurrence Matrix $X$
To obtain the concurrent number of likes $X_{i,j}$ we must first binary encode (`0` or `1`) each `"rating"` in the `ratings` DataFrame. Let us encode the value of liking a movie for each review as such

$$ \text{Liked}(\text{Rating}) =
    \begin{cases}
        1 & \text{if Rating}\geq 4\\
        0 & \text{otherwise}
    \end{cases}$$

and store these values in a new `"liked"` column. We can subsequently drop the unnecessary `rating` and `timestamp` columns after this process.

In [4]:
import numpy as np

# create liked column
ratings["liked"] = np.where(ratings["rating"] >= 4, 1, 0)

# drop columns
ratings.drop(["rating", "timestamp"], axis=1, inplace=True)

We next create the `movie_ratings` DataFrame by joing the `movies` and `ratings` DataFrames. Setting the `merge` method parameter `how="left"` ensures that the original number of movies, 9,742, are maintained after the join.

In [5]:
# left join on movieId
movie_ratings = pd.merge(movies, ratings, how="left", on="movieId").reset_index()

A user-likes interaction matrix can then be constructed using the `pivot_table` method, whose rows correspond to the number of unique users `n_users` and columns correspond to the number of unique movies `n_movies` from the original data. This results in a sparse matrix whose rows summarize each users liked movies.

In [24]:
# pivot table on userId
user_likes = movie_ratings.pivot_table(values="liked", index="userId", columns="movieId", dropna=False, fill_value=0)

user_likes

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1.0           1       0       1       0       0       1       0       0   
2.0           0       0       0       0       0       0       0       0   
3.0           0       0       0       0       0       0       0       0   
4.0           0       0       0       0       0       0       0       0   
5.0           1       0       0       0       0       0       0       0   
...         ...     ...     ...     ...     ...     ...     ...     ...   
606.0         0       0       0       0       0       0       0       0   
607.0         1       0       0       0       0       0       0       0   
608.0         0       0       0       0       0       0       0       0   
609.0         0       0       0       0       0       0       0       0   
610.0         1       0       0       0       0       1       0       0   

movieId  9       10      ...  193565  193567  193571  193573  193579  193581  \
userId                   ...                                                   
1.0           0       0  ...       0       0       0       0       0       0   
2.0           0       0  ...       0       0       0       0       0       0   
3.0           0       0  ...       0       0       0       0       0       0   
4.0           0       0  ...       0       0       0       0       0       0   
5.0           0       0  ...       0       0       0       0       0       0   
...         ...     ...  ...     ...     ...     ...     ...     ...     ...   
606.0         0       0  ...       0       0       0       0       0       0   
607.0         0       0  ...       0       0       0       0       0       0   
608.0         0       1  ...       0       0       0       0       0       0   
609.0         0       1  ...       0       0       0       0       0       0   
610.0         0       0  ...       0       0       0       0       0       0   

movieId  193583  193585  193587  193609  
userId                                   
1.0           0       0       0       0  
2.0           0       0       0       0  
3.0           0       0       0       0  
4.0           0       0       0       0  
5.0           0       0       0       0  
...         ...     ...     ...     ...  
606.0         0       0       0       0  
607.0         0       0       0       0  
608.0         0       0       0       0  
609.0         0       0       0       0  
610.0         0       0       0       0  

[610 rows x 9742 columns]

Co-occurrence matrix $X$ can now be constructed as the inner (dot) product of the transpose of `user_likes ` and itself. Element $X_{ij} \in X$ corresponds to the number of users that like both movie $i$ and $j$.

In [26]:
# convert to numpy ndarray for dot product computation
user_likes_array = user_likes.to_numpy()

X_co_occurrence = np.dot(user_likes_array.T, user_likes_array)


In [28]:
# display as DataFrame for clarity
X_display = pd.DataFrame(X_co_occurrence, index=movies.movieId, columns = movies.movieId)\
    
X_display

movieId  1       2       3       4       5       6       7       8       \
movieId                                                                   
1           147      21      11       0       7      27       7       1   
2            21      50       5       0       4       8       6       0   
3            11       5      18       0       4       4       5       1   
4             0       0       0       0       0       0       0       0   
5             7       4       4       0      12       3       4       1   
...         ...     ...     ...     ...     ...     ...     ...     ...   
193581        0       0       0       0       0       0       0       0   
193583        0       0       0       0       0       0       0       0   
193585        0       0       0       0       0       0       0       0   
193587        0       0       0       0       0       0       0       0   
193609        0       0       0       0       0       0       0       0   

movieId  9       10      ...  193565  193567  193571  193573  193579  193581  \
movieId                  ...                                                   
1             4      19  ...       0       0       0       0       0       0   
2             0       9  ...       0       0       0       0       0       0   
3             2       3  ...       0       0       0       0       0       0   
4             0       0  ...       0       0       0       0       0       0   
5             1       2  ...       0       0       0       0       0       0   
...         ...     ...  ...     ...     ...     ...     ...     ...     ...   
193581        0       0  ...       0       0       1       1       0       1   
193583        0       0  ...       0       0       0       0       0       0   
193585        0       0  ...       0       0       0       0       0       0   
193587        0       0  ...       0       0       0       0       0       0   
193609        0       0  ...       0       0       0       0       0       0   

movieId  193583  193585  193587  193609  
movieId                                  
1             0       0       0       0  
2             0       0       0       0  
3             0       0       0       0  
4             0       0       0       0  
5             0       0       0       0  
...         ...     ...     ...     ...  
193581        0       0       0       0  
193583        0       0       0       0  
193585        0       0       0       0  
193587        0       0       0       0  
193609        0       0       0       1  

[9742 rows x 9742 columns]